independent-projects/qute/core/src/main/java/io/quarkus/qute/ReflectionValueResolver.java

```diff
@@ -0,0 +1,175 @@
+package io.quarkus.qute;
+
+import java.lang.reflect.Array;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class ReflectionValueResolver implements ValueResolver {
+
+    /**
+     * Lazy loading cache of lookup attempts (contains both hits and misses)
+     */
+    private final ConcurrentMap<MemberKey, Optional<MemberWrapper>> memberCache = new ConcurrentHashMap<>();
+
+    private static final MemberWrapper ARRAY_GET_LENGTH = Array::getLength;
+
+    public static final String GET_PREFIX = "get";
+    public static final String IS_PREFIX = "is";
+
+    @Override
+    public int getPriority() {
+        return -1;
+    }
+
+    @Override
+    public boolean appliesTo(EvalContext context) {
+        return context.getBase() != null;
+    }
+
+    @Override
+    public CompletionStage<Object> resolve(EvalContext context) {
+
+        Object base = context.getBase();
+        MemberKey key = MemberKey.newInstance(base, context.getName());
+        MemberWrapper wrapper = memberCache.computeIfAbsent(key, ReflectionValueResolver::findWrapper).orElse(null);
```

Comment:
Sanne says that if you expect a lot of hits, it's more performant to do a proper `get` before that.

Reply:
1) reflection resolver is just a fallback, 2) `get()` without sync would not be atomic